[RFC 140] Simple package paths, part 1b: Enabling the directory structure (#237439)
NixOS/nixpkgs: pkgs/test/by-name/check-references/README.md

> Usage: nix-reference-check [FILES]

Reviewer: ```suggestion
Usage: check-references [DIR]
```

At least when I built it, the binary that got built was called `check-references`.

Also, it insists on being passed a directory; you can't pass it just a single `.nix` file.

Author: This is outdated, the check is now in https://github.com/NixOS/nixpkgs/pull/250885, called `nixpkgs-check-by-name <NIXPKGS>`, and there's no support for files (it takes a Nixpkgs path now).

> Checks that all given Nix files don't transitively and statically refer to files outside their own directory

Reviewer: ```suggestion
Checks that all given Nix files don't statically refer to files outside their own directory
```

> Checks that all given Nix files don't transitively and statically refer to files outside their own directory

Reviewer: ```suggestion
Checks that all given Nix files don't transitively and statically refer to files outside the directory which contains them.

Dynamic references, produced by using `+` to concatenate a path with a string, are not checked.  For example, this tool will not prevent:

  import (./. + "/../../../default.nix")
```

Reviewer: I tried this test case and `check-references` allowed it.  IIUC this is intended, since the tool only does what can be done without having to evaluate anything.

Author: Oh yeah that's a good point. I'm not sure to what extent these exact semantics should be specified though. It's not really important for contributors to know the exact rules, and if they're interested, they could check the code or try it out. And I'd hope that reviewers catch people doing such workarounds.
